# Assignment: regression

In [6]:
import graphlab
import pandas as pd
import numpy as np

## 1. Selection and summary statistics

 In the notebook we covered in the module, we discovered which neighborhood (zip code) of Seattle had the highest average house sale price. Now, take the sales data, select only the houses with this zip code, and compute the average price.

In [4]:
sales = graphlab.SFrame('home_data.gl/')

In [24]:
# El zipcode era 98039

sales_98039 = sales[sales['zipcode'] == '98039']
print sales_98039['price'].mean()

2160606.6


**Respuesta:** el precio medio de ese zipcode es `2160606.6`.

## 2. Filtering data

One of the key features we used in our model was the number of square feet of living space (‘sqft_living’) in the house. For this part, we are going to use the idea of filtering (selecting) data.

In particular, we are going to use logical filters to select rows of an SFrame. You can find more info in the Logical Filter section of this documentation. Using such filters, first select the houses that have ‘sqft_living’ higher than 2000 sqft but no larger than 4000 sqft.What fraction of the all houses have ‘sqft_living’ in this range? Save this result to answer the quiz at the end.

In [34]:
# Filtramos

sales_2_4 = sales[
    (sales['sqft_living'] > 2000) & (sales['sqft_living'] < 4000)]

In [65]:
# Vemos el número de filas para ambas tablas

print len(sales_2_4)
print len(sales)

9111
21613


In [71]:
frac = len(sales_2_4)/len(sales)
frac

0

## 3. Building a regression model with several more features: 

In the sample notebook, we built two regression models to predict house prices, one using just ‘sqft_living’ and the other one using a few more features, we called this set

In [51]:
my_features = ['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'zipcode']

 Now, going back to the original dataset, you will build a model using the following features:

In [52]:
advanced_features = [
'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'zipcode',
'condition', # condition of house				
'grade', # measure of quality of construction				
'waterfront', # waterfront property				
'view', # type of view				
'sqft_above', # square feet above ground				
'sqft_basement', # square feet in basement				
'yr_built', # the year built				
'yr_renovated', # the year renovated				
'lat', 'long', # the lat-long of the parcel				
'sqft_living15', # average sq.ft. of 15 nearest neighbors 				
'sqft_lot15', # average lot size of 15 nearest neighbors 
]

Note 1: both models must be trained on the original sales dataset, not the filtered one.

Note 2: when doing the train-test split, make sure you use seed=0, so you get the same training and test sets, and thus results, as we do.

Note 3: in the module we discussed residual sum of squares (RSS) as an error metric for regression, but GraphLab Create uses root mean squared error (RMSE). These are two common measures of error regression, and RMSE is simply the square root of the mean RSS:

where N is the number of data points. RMSE can be more intuitive than RSS, since its units are the same as that of the target column in the data, in our case the unit is dollars ($), and doesn't grow with the number of data points, like the RSS does.

(Important note: when answering the question below using GraphLab Create, when you call the linear_regression.create() function, make sure you use the parameter validation_set=None, as done in the notebook you download above. When you use regression GraphLab Create, it sets aside a small random subset of the data to validate some parameters. This process can cause fluctuations in the final RMSE, so we will avoid it to make sure everyone gets the same answer.)

**What is the difference in RMSE between the model trained with my_features and the one trained with advanced_features? Save this result to answer the quiz at the end.**

In [54]:
# Dividimos en train y test set

train_data,test_data = sales.random_split(.8,seed=0)

In [55]:
myfeat_model = graphlab.linear_regression.create(train_data, target='price', features=my_features,validation_set=None)

Linear regression:

--------------------------------------------------------

Number of examples          : 17384

Number of features          : 6

Number of unpacked features : 6

Number of coefficients    : 115

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+--------------------+---------------+

| Iteration | Passes   | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+--------------+--------------------+---------------+

| 1         | 2        | 1.074187     | 3763208.270524     | 181908.848367 |

+-----------+----------+--------------+--------------------+---------------+

SUCCESS: Optimal solution found.

In [56]:
adv_model = graphlab.linear_regression.create(train_data, target='price', features=advanced_features,validation_set=None)

Linear regression:

--------------------------------------------------------

Number of examples          : 17384

Number of features          : 18

Number of unpacked features : 18

Number of coefficients    : 127

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+--------------------+---------------+

| Iteration | Passes   | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+--------------+--------------------+---------------+

| 1         | 2        | 0.148951     | 3469012.450487     | 154580.940732 |

| 2         | 3        | 0.247205     | 3469012.450673     | 154580.940735 |

+-----------+----------+--------------+--------------------+---------------+

SUCCESS: Optimal solution found.

In [60]:
# Calculamos la diferencia de RMSE

rmse_myf = myfeat_model.evaluate(test_data)['rmse']
rmse_adv = adv_model.evaluate(test_data)['rmse']

In [61]:
print rmse_myf
print rmse_adv

179542.433313
156831.116802


In [63]:
# Diferencia

rmse_myf - rmse_adv

22711.316510682926